### PR TITLE
Adds OpenAI functions powered document metadata tagger

### PR DESCRIPTION
**Pull request by jacoblee93**
Adds a new document transformer that automatically extracts metadata for a document based on an input schema. I also moved `document_transformers.py` to `document_transformers/__init__.py` to group it with this new transformer - it didn't seem to cause issues in the notebook, but let me know if I've done something wrong there.

Also had a linter issue I couldn't figure out:

```
MacBook-Pro:langchain jacoblee$ make lint
poetry run mypy .
docs/dist/conf.py: error: Duplicate module named "conf" (also at "./docs/api_reference/conf.py")
docs/dist/conf.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
docs/dist/conf.py: note: Common resolutions include: a) using `--exclude` to avoid checking one of them, b) adding `__init__.py` somewhere, c) using `--explicit-package-bases` or adjusting MYPYPATH
Found 1 error in 1 file (errors prevented further checking)
make: *** [lint] Error 2
```

@rlancemartin @baskaryan

### CODE REVIEW

**Comment by bitomukesh** [comment]
@CodiumAI-Agent

File: tests/integration_tests/vectorstores/test_pinecone.py [hunk 1]
```diff
@@ -2,9 +2,9 @@
 import os
 import time
 import uuid
-import numpy as np
```
Comment (jacoblee93) [inline tests/integration_tests/vectorstores/test_pinecone.py:5]: Didn't change these manually, just ran `make format`.

File: langchain/document_transformers/openai_functions.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    def transform_documents(
+        self, documents: Sequence[Document], **kwargs: Any
+    ) -> Sequence[Document]:
+        """Automatically extract and populate metadata
+        for each document according to the provided schema."""
+
+        new_documents = []
+
+        for document in documents:
+            extracted_metadata = self.tagging_chain.__call__(document.page_content)
```
Comment (baskaryan) [inline langchain/document_transformers/openai_functions.py:26]: can just self.tagging_chain(document.page_content)

**Review by baskaryan — APPROVED** [review]
lgtm, just one nit

File: langchain/document_transformers/__init__.py
Comment (baskaryan) [inline langchain/document_transformers/__init__.py:1]: we should probably move the stuff here to their own files. think #7379 already doing that so will wait till it lands to merge

**Comment by rlancemartin** [comment]
Nice. Yes. This is good. Consolidating `document_transformers` in a new dir will make additions easier going forward.

Like @baskaryan said, https://github.com/hwchase17/langchain/pull/7379 is going in soon.

**Pull request merged**